File: cpp/factorielle.cpp
#include <iostream>
#include <iomanip>

int factorielle_recursive(int n) 
{
    switch (n) 
    {
        case 0: case 1:
            return 1;
        default:
            return n * factorielle_recursive(n - 1);
    }
}


int factorielle_recursive_avec_si(int n) 
{
    if (n == 0 || n == 1)
        return 1;
    return n * factorielle_recursive_avec_si(n - 1);
}

int factorielle_iterative(int n) 
{
    int resultat = 1;
    while (n != 0) 
    {
        resultat *= n;
        n -= 1;
    }
    return resultat;
}

int main() {
    for (int i = 0; i < 10; ++i) 
    {
        std::cout << "factorielle(" 
                  << std::setw(2) << i << ") = "
                  << std::setw(7) << std::right << factorielle_recursive(i) << " (recursif) "
                  << std::setw(7) << std::left  << factorielle_iterative(i) << " (iteratif)"
                  << std::endl;
    }
    return 0;
}

/*
g++ -o cpp/factorielle cpp/factorielle.cpp
./cpp/factorielle
*/

File: rust/src/factorielle.rs
pub const fn factorielle_recursive(n : u32) -> u32
{
    match n
    {
        0 | 1 => 1,
        _ => n * factorielle_recursive(n-1)
    }
}

pub const fn factorielle_recursive_avec_si(n : u32) -> u32
{
    if n == 0 || n == 1 { 1 } 
    else { n * factorielle_recursive(n-1) }
}

pub const fn factorielle_iterative(mut n : u32) -> u32
{
    let mut resultat = 1;
    while n != 0
    {
        resultat *= n;
        n -= 1;
    }
    resultat
}

pub fn factorielle_exemple()
{
    let mut v = Vec::new();

    for i in 0..10
    {
        println!("factorielle({: >2}) = {: >7} (recursif) {: <7} (iteratif)", i, factorielle_recursive(i), factorielle_iterative(i))
    }
    println!();
}

/* 
.\rust\target\debug\jpo.exe
*/